### PR TITLE
Surface errors when deleting models

### LINKED ISSUE #3732: Errors destroying models aren't surfaced to user
On staging, there's an intermittent bug with juju unrelated to the GUI that prevents destroying a model:

```
{"type":"ModelManager","request":"DestroyModels","params":{"models":[{"model-tag":"model-31fea4e9-36eb-41c7-82f3-0cae8e6e259b"}]},"version":4,"request-id":8}
{"request-id":8,"response":{"results":[{"error":{"message":"this Juju controller requires destroyStorage to be true","code":""}}]}}
```

When destroying this model in the GUI, no message is shown to the user, and it should be.

### PR DESCRIPTION
**Pull request by makyo**
Fixes #3732

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8909/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8910/

File: jujugui/static/gui/src/app/components/profile/model-list/model-list.js [hunk 1]
```diff
@@ -74,6 +74,29 @@ class ProfileModelList extends React.Component {
     });
   }
 
+  _confirmDestroy(modelUUID) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/profile/model-list/model-list.js]: docblock missing.

File: jujugui/static/gui/src/app/components/profile/model-list/model-list.js [hunk 2]
```diff
@@ -74,6 +74,29 @@ class ProfileModelList extends React.Component {
...
+    this.setState({notification: null});
+    this.props.destroyModel(modelUUID, (error, data) => {
+      if (error) {
+        this.props.addNotification({
+          title: 'Error destroying model',
+          message: error,
+          level: 'error'
+        });
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/profile/model-list/model-list.js]: `return;` after this msg else you could give two notifications for the same error....or maybe not....see comment below about multiple error locations.

File: jujugui/static/gui/src/app/components/profile/model-list/model-list.js [hunk 3]
```diff
@@ -74,6 +74,29 @@ class ProfileModelList extends React.Component {
...
+      }
+      if (data != {}) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/profile/model-list/model-list.js]: I believe here you're checking to see if the data object returned is not empty. I'm a little confused here as to why an error destroying model wouldn't be returned as an error above. If the `destroyModel` call returns error values in the data then maybe we need to document that, or fix it in the api.

In any event, if you want to check if an object is empty you should use something like `if (Object.keys(data).length) {` as it'll only report keys for enumerable properties.

Comment (makyo) [reply jujugui/static/gui/src/app/components/profile/model-list/model-list.js]: You can have an error from the API call, or models can return an error on destroy (remember that the actual call is destroyModels and takes more than one; the object is a map of model uuids to responses), or potentially both, so we need to display all errors. I'll comment to that effect.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8911/

**Comment by makyo** [comment]
# QA

Deploy http://localhost:8042/canonical-kubernetes/bundle/76

Attempt to destroy the model once deployment is finished. Ensure that you get an error about storage.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8913/

File: jujugui/static/gui/src/app/store/env/controller-api.js [hunk 4]
```diff
@@ -856,15 +856,21 @@ window.yui.add('juju-controller-api', function(Y) {
...
+          if (result.error) {
+            prev[id] = result.error.message;
+            errors.push(result.error.message);
+          } else {
+            prev[id] = null;
+          }
           return prev;
         }, {});
-        callback(null, results);
+        callback(errors, results);
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/controller-api.js]: `errors` here would be truthy (`[]` is truthy) so I'd prefer it to stay as null so consumers of the api could still go `if (!error) { ... }`

What do you think about: 
`callback(errors.length ? errors : null, results);`

Comment (makyo) [reply jujugui/static/gui/src/app/store/env/controller-api.js]: This is such a ridiculously small win :P If you say so...

Comment (hatched) [reply jujugui/static/gui/src/app/store/env/controller-api.js]: It's just about creating a consistent api for those who are consuming it.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8914/

**Comment by hatched** [comment]
Thanks @makyo :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**